### PR TITLE
feat: Add prefix to all hash class to avoid collision when needed

### PR DESCRIPTION
**Pull request by rafael-cineis-ifood**
## What?
Created an environment variable `SC_CLASS_PREFIX` to add a prefix to all hash classes generated, including the ones that contain the style itself to avoid unexpected class collision.

## Why?
To avoid classes collision when we have multiple styled-components loaded at the same time, usually it happens when working with a micro-frontend structure and we have 2 micro-frontend loading the styled-components.

There is the option to use the offical babel plugin to add the [namespace option](https://styled-components.com/docs/tooling#namespace), but as the plugin only add the prefix in the `sc-???` class, it will not shield us from class collision between libraries used in common by multiple micro-frontends because usually the other classes (the ones with the style itself) are the ones that collides.

Based on the description above and the issue described here: https://github.com/styled-components/babel-plugin-styled-components/issues/308, I would like to propose an option to add a prefix to those classes that contain the css and with that shield the styled-components generated classes from colliding with something external when we set this variable.

## How?
Update the function responsible to generate all the class hashes and concatenate the env variable, this variable would be possible to be set by `REACT_APP_SC_CLASS_PREFIX` or `SC_CLASS_PREFIX`, if they aren't set we will use and empty string which will not affect anything.

## Testing
Change the file `packages/sandbox/server/webpack.config.js` by adding the option `SC_CLASS_PREFIX: 'custom-class-prefix_',` inside the webpack EnvironmentPlugin.

### Before
image

### After
By adding the environment variable:
```javascript
new webpack.EnvironmentPlugin({
  // This will provide default value for NODE_ENV, unless defined otherwise
  // more info: https://webpack.js.org/plugins/environment-plugin/#usage-with-default-values
  NODE_ENV: 'development',
  SC_CLASS_PREFIX: 'custom-class-prefix_',
}),
```

And running the sandbox, we can see the hash classes with the prefix provided in the env variable:
image

## References
Special thanks to @justnewbee for contributing it in https://github.com/styled-components/styled-components/pull/3225, I'm trying to reopen this issue with more tests and talking about why the babel plugin isn't working for us.

### CODE REVIEW

File: packages/styled-components/src/utils/test/generateAlphabeticName.test.js
Comment (guimello) [inline packages/styled-components/src/utils/test/generateAlphabeticName.test.js:27]: I guess this is a development leftover, `only`.
👀

Comment (rafael-cineis-ifood) [reply packages/styled-components/src/utils/test/generateAlphabeticName.test.js]: Oh, my bad, thanks for the review :smile:

**Comment by kassi** [comment]
Hey @rafael-cineis-ifood  and @guimello : What's the status on this?
We would like to see this being merged since we would like to use it#s functionality rather than having some hacks doing something similar in our environment only.
Anything we can do to push this forward?

**Comment by guimello** [comment]
> Hey @rafael-cineis-ifood and @guimello : What's the status on this?
> We would like to see this being merged since we would like to use it's functionality rather than having some hacks doing something similar in our environment only.
> Anything we can do to push this forward?

I think we are just waiting for maintainers to weigh in

**Comment by rafael-cineis-ifood** [comment]
> Hey @rafael-cineis-ifood and @guimello : What's the status on this?
> We would like to see this being merged since we would like to use it's functionality rather than having some hacks doing something similar in our environment only.
> Anything we can do to push this forward?

As @guimello mentioned, we are just waiting for the review from any maintainers, I already tried contacting them but no response. We are creating a fork right now to handle it for our micro-frontends.

**Comment by kassi** [comment]
Ok, thanks for the update. Sounds reasonable. We may be using it as well in the same way then.

**Comment by rafael-cineis-ifood** [comment]
> Ok, thanks for the update. Sounds reasonable. We may be using it as well in the same way then.
> Do you plan a public fork or private one?

Right now we are going with a private fork with hope that this feature got reviewed and somehow handled properly inside styled-components official solution, we don't want to keep this fork for too long or we will need to start to keep it up-to-date and it could be a pain.

**Review by taranek — APPROVED** [review]

**Comment by c0d3-k1ra** [comment]
Hey there,

Any idea when this can be merged? We are also facing the same issue where styles have collision.

**Comment by rafael-cineis-ifood** [comment]
Hi @probablyup, I saw that you reviewed some other PRs related to this theme, could you have a look at it please?

**Review by kbalus — APPROVED** [review]

**Comment by kitten** [comment]
I'm currently inclined not to merge this unfortunately. There are several other options that can be combined to combat conflicts, in case the library isn't shared/federated in an app, which I've written down here: https://github.com/styled-components/styled-components/issues/2974#issuecomment-895467259

> However, this is a hash based on component IDs and the contents of the CSS string, so collisions will be extremely unlikely given the namespace option, read: as unlikely as any other hash collision with 32 bits if no mistake is made in our hashing logic.

Technically, the hash classes that are used for styling hence shouldn't conflict. If they do conflict (which is something I think we had once for a past release, then their entryopy is likely bugged), then it's more likely to be something that needs to be fixed in the library. In other words, it'd be mostly a visual change at the cost of adding yet another environment variable.

To also mention the methods to actually do address micro frontends that aren't federating styled-components as a dependency:
https://github.com/styled-components/styled-components/issues/2974#issuecomment-895301213

> So, I will stand by this, but either you share styled-components like many microfrontend builds share react/react-dom across a given page.
> Or, you can use multiple versions of styled-components on the same page by altering process.env.SC_ATTR per build. I haven't tested this in a while, but it should still work just fine.
> Also, to summarise, the babel-plugin namespace option can now also be used to prevent hash collisions: https://styled-components.com/docs/tooling#namespace
> Since in all cases a combination of these three solutions work, I'll only leave this issue open for a bit and will lock it after, since it's gotten too long. I appreciate if any other issues are highlighted here, but I'm fairly certain at this point that hunting for "the big fix" isn't really a replacement for the above.

But what's going to be the most likely option here is using `namespace`, potentially with `SC_ATTR`, and the hash-based classes shouldn't conflict.

File: packages/styled-components/src/utils/generateAlphabeticName.js
Comment (jonasantonelli) [inline packages/styled-components/src/utils/generateAlphabeticName.js:26]:
```suggestion
  const prefix = SC_CLASS_PREFIX ? `${SC_CLASS_PREFIX}-` : '';
```

What do you think of using a hyphen, like it's done with the "sc-" prefix?
`sc-pFZIQ`

**Comment by rafael-cineis-ifood** [comment]
> I'm currently inclined not to merge this unfortunately. There are several other options that can be combined to combat conflicts, in case the library isn't shared/federated in an app, which I've written down here: [#2974 (comment)](https://github.com/styled-components/styled-components/issues/2974#issuecomment-895467259)
> 
> > However, this is a hash based on component IDs and the contents of the CSS string, so collisions will be extremely unlikely given the namespace option, read: as unlikely as any other hash collision with 32 bits if no mistake is made in our hashing logic.
> 
> Technically, the hash classes that are used for styling hence shouldn't conflict. If they do conflict (which is something I think we had once for a past release, then their entryopy is likely bugged), then it's more likely to be something that needs to be fixed in the library. In other words, it'd be mostly a visual change at the cost of adding yet another environment variable.
> 
> To also mention the methods to actually do address micro frontends that aren't federating styled-components as a dependency:
> [#2974 (comment)](https://github.com/styled-components/styled-components/issues/2974#issuecomment-895301213)
> 
> > So, I will stand by this, but either you share styled-components like many microfrontend builds share react/react-dom across a given page.
> > Or, you can use multiple versions of styled-components on the same page by altering process.env.SC_ATTR per build. I haven't tested this in a while, but it should still work just fine.
> > Also, to summarise, the babel-plugin namespace option can now also be used to prevent hash collisions: https://styled-components.com/docs/tooling#namespace
> > Since in all cases a combination of these three solutions work, I'll only leave this issue open for a bit and will lock it after, since it's gotten too long. I appreciate if any other issues are highlighted here, but I'm fairly certain at this point that hunting for "the big fix" isn't really a replacement for the above.
> 
> But what's going to be the most likely option here is using `namespace`, potentially with `SC_ATTR`, and the hash-based classes shouldn't conflict.

Hi @kitten , sorry for the late response... I tried using the `namespace` only as also the `namespace + SC_ATTR` but the hashes continue conflicting, if I click on the classes I could see that it was inside the `<script />` tag with the `SC_ATTR` from each application, but as they were loaded in the same container application, it caused this issue.

But I agree that probably the best idea would be to have the styled-components shared between the applications via a CDN or module federation.

File: packages/styled-components/src/utils/generateAlphabeticName.js
Comment (rafael-cineis-ifood) [reply packages/styled-components/src/utils/generateAlphabeticName.js:26]: It's following the same pattern that the [namespace option](https://styled-components.com/docs/tooling#namespace) is using, it adds the prefix with an underscore after it

Comment (jonasantonelli) [reply packages/styled-components/src/utils/generateAlphabeticName.js:26]: Great, make sense for me

**Pull request closed without merging**
